src/main/java/org/dependencytrack/tasks/metrics/VulnerabilityMetricsUpdateTask.java

```diff
@@ -59,104 +62,75 @@ private void updateMetrics() throws Exception {
-                    }
-                }
-
-                LOGGER.debug("Fetching next vulnerabilities page");
-                final long lastId = vulnerabilities.get(vulnerabilities.size() - 1).getId();
-                vulnerabilities = fetchNextVulnerabilitiesPage(pm, lastId);
-            }
-
-            for (final VulnerabilityMetrics metric : yearMonthCounters.getMetrics()) {
-                qm.synchronizeVulnerabilityMetrics(metric);
-            }
-            for (final VulnerabilityMetrics metric : yearCounters.getMetrics()) {
-                qm.synchronizeVulnerabilityMetrics(metric);
-            }
-        }
-
-        LOGGER.info("Completed metrics update on vulnerability database in " +
-                DurationFormatUtils.formatDuration(new Date().getTime() - measuredAt.getTime(), "mm:ss:SS"));
-    }
-
-    /**
-     * Fetch {@link Vulnerability}s in pages of {@code 500}.
-     *
-     * @param pm     The {@link  PersistenceManager} to use
-     * @param lastId Highest ID of the previously fetched page
-     * @return Up to {@code 500} {@link Vulnerability} objects
-     * @throws Exception If the query could not be closed
-     */
-    private List<Vulnerability> fetchNextVulnerabilitiesPage(final PersistenceManager pm, final Long lastId) throws Exception {
-        try (final Query<Vulnerability> query = pm.newQuery(Vulnerability.class)) {
-            if (lastId != null) {
-                query.setFilter("id < :lastId");
-                query.setParameters(lastId);
-            }
-            query.setOrdering("id DESC");
-            query.range(0, 500);
-            query.getFetchPlan().setGroup(Vulnerability.FetchGroup.METRICS_UPDATE.name());
-            return List.copyOf(query.executeList());
+            /**
+             * 
+             * The created field has priotiy over the published field, which is used as a fallback
+             * However, the created field is always empty (in my instances)
+             * But we leave this mechanism and field juggling in place for backwards compatibility,
+             * and for (future) analyzers/sources that might provide this field.
+             * 
+             * BTW the queries to get these vulnerability counts are very fast,
+             * so strictly speaking there is no reason to create this extra table with metrics
+             *  
+             */
+
+            // Get metrics by published date but only if created field is null
+            Collection<YearMonthMetric> published = queryForMetrics(pm, "published", "created");
+
+            // Get metrics by created date regardless of value of published field
+            Collection<YearMonthMetric> created = queryForMetrics(pm, "created", null);
+
+            // Merge flat lists
+            published.addAll(created);
+
+            // Collect into nested map so we can sum the counts
+            Map<Object, Map<Object, Long>> metrics = published.stream()
```

Comment:
If you provide concrete type arguments here, you can avoid the casting to `Integer` later on:

```suggestion
            Map<Integer, Map<Integer, Long>> metrics = published.stream()
```

Reply:
ok thanks, in one of the earlier iterations it wouldn't accept this cast